Higher Performance nuclinfo

ALescoulie opened the pull request:
Fixes #3600, #3310

Changes made in this Pull Request:
 - Add `AnalysisBase` derived classes for calculating different nucleic acid base pair distances.

I built a new class `NucPairDist` that calculates distances between atoms in nucleic acid. It has 3 subclasses `WCDist`, `MajorDist` and `MinorDist`. Those are indented to  improve the performance of their equivalents in nucinfo. They also support taking groups of base pairs and running the distance over a trajectory.

TODO:
- [x] Docs

PR Checklist
------------
 - [x] Tests?
 - [x] Docs?
 - [x] CHANGELOG updated?
 - [x] Issue raised/referenced?

ALescoulie committed:
commit WCBase

ALescoulie committed:
commit nucleicacids tests

ALescoulie committed:
commit nucleicaid

ALescoulie committed:
commit NucPairDist

ALescoulie committed:
finish WCDist

ALescoulie committed:
commit wc_dist test

ALescoulie committed:
commit major and minor dist

ALescoulie committed:
add tests

pep8speaks commented:
Hello @ALescoulie! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:



* In the file [`package/MDAnalysis/analysis/nucleicacids.py`](https://github.com/MDAnalysis/mdanalysis/blob/374ba4927c34ee0f156e92d039267ecbc538cc63/package/MDAnalysis/analysis/nucleicacids.py):

> [Line 105:80](https://github.com/MDAnalysis/mdanalysis/blob/374ba4927c34ee0f156e92d039267ecbc538cc63/package/MDAnalysis/analysis/nucleicacids.py#L105): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (86 > 79 characters)
> [Line 141:80](https://github.com/MDAnalysis/mdanalysis/blob/374ba4927c34ee0f156e92d039267ecbc538cc63/package/MDAnalysis/analysis/nucleicacids.py#L141): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (86 > 79 characters)
> [Line 142:80](https://github.com/MDAnalysis/mdanalysis/blob/374ba4927c34ee0f156e92d039267ecbc538cc63/package/MDAnalysis/analysis/nucleicacids.py#L142): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (84 > 79 characters)





##### Comment last updated at 2022-04-19 08:17:12 UTC

ALescoulie commented:
Just realized that pandas is not included in MDAnalysis, I had been using tidy DataFrames to record my results

IAlibay commented:
I'll let the other @MDAnalysis/coredevs speak here too, but I'm not particularly keen on having pandas as a core dependency (in fact my current goal is to remove more dependencies from MDAnalysis). I would prefer if numpy arrays could be used to store results where possible.

ALescoulie commented:
@IAlibay Don't think switching to arrays will be too much trouble, I'm trying to store selection, distance and time results in an organized way so I deflated to a DataFrame.

orbeckst commented:
I second @IAlibay 's https://github.com/MDAnalysis/mdanalysis/pull/3611#issuecomment-1092257941 — keep it as a numpy array. I don't see pandas here to be essential. 

If you _really_ need to make data available then you could create `to_df()` that imports pandas only in the function and that can fail and tell you to install it. We have packages that are optional and then we follow that approach.

orbeckst commented:
Btw, pandas.DataFrame does not have great performance in my experience — you get convenience but not the speed of numpy arrays.

ALescoulie commented:
@orbeckst I don't use it during iteration, I just build it from a dictionary in the `_conclude` step, but I could just return a names list and a numpy array of distances and times. Overall is this like what you were wanting for #3600 .

orbeckst commented:
I understand; my point was that when you want to process your results further (e.g., calculate statistics) then this will be faster when the data are already an array than if it were a DataFrame. For people comfortable with pandas, the df is probably more convenient and organized. MDAnalysis has been embracing numpy arrays as the foundation for everything so it makes sense to use them for results, too. But I don't see anything wrong with adding a convenience function that presents the raw data as a DataFrame (although other devs may have different opinions on the matter).

ALescoulie commented:
@orbeckst and @IAlibay  looking at how my DataFrame (selection, time, dist as cols with results in rows) is structured, how do you think is best to structure the results into an array. I'm thinking a dictionary with selection number as keys.

orbeckst commented:
I'd say look at some of the other standard analysis, e.g., RMSD.

Also @PicoCentauri and @joaomcteixeira might also have an opinion on how to store data in `results`.

PicoCentauri commented:
I am also in favor of keeping things as `numpy` arrays. I don't see much benefit of a pandas data frame here. Additionally, the new classes will be part of the [CLI](https://github.com/MDAnalysis/mdacli), and the CLI's saving workflow is (currently) not able to handle data frames. The latter is maybe an issue of the CLI I have to fix though.

@ALescoulie I would also be very happy if you could test your modules from the command (once you have a proper docstring)!

orbeckst commented on package/MDAnalysis/analysis/nucleicacids.py:
```diff
@@ -0,0 +1,187 @@
+
+import numpy as np
+import pandas as pd
+from math import pi, sin, cos, atan2, sqrt, pow
```
Whenever I see `import math` I fear that there's code that could be made faster by using numpy array operations...

ALescoulie replied:
I'll use the numpy implementations, that was just copied from nucinfo.py I haven't implemented the other half of it using those operations yet

orbeckst commented on package/MDAnalysis/analysis/nucleicacids.py:
```diff
@@ -0,0 +1,187 @@
+import MDAnalysis as mda
+from MDAnalysis.lib import mdamath
+from MDAnalysis.analysis.base import AnalysisBase
+from package import MDAnalysis
+from package.MDAnalysis.exceptions import SelectionError
```
There's no package module. This shouldn't even work.... except if you're running from the top of the checked out git repo.

If you want to do development, install with
```
pip install -e package/
```

ALescoulie replied:
I got the imports fixed not sure why it was like that

orbeckst commented on package/MDAnalysis/analysis/nucleicacids.py:
```diff
@@ -0,0 +1,187 @@
+    """A named tuple for storing the selection information of a specific base pair of a nucleic acid.
+    This makes passing in selections more organized.
+
+    :Arguments:
+    *segid*
+        The name of the segment the base is on
+
+    *base*
+        The number of the base pair in the segment
+
+    .. rubric:: Example
```
We use numpydoc formatting for doc strings. See https://userguide.mdanalysis.org/stable/contributing_code.html#working-with-the-code-documentation

ALescoulie replied:
Ok, I'll start on the docs soon now that I fixed the some of the functionality of the code.

orbeckst commented on package/MDAnalysis/analysis/nucleicacids.py:
```diff
@@ -0,0 +1,187 @@
+    _s2: mda.AtomGroup
+    _res_dict: Dict[str, List[Union[float, str]]]
+    _names: List[str]
+    results: pd.DataFrame
```
results has to be a [Results](https://docs.mdanalysis.org/stable/documentation_pages/analysis/base.html#MDAnalysis.analysis.base.Results) container to be compatible with AnalysisBase/mdacli

ALescoulie replied:
I'm switching over to that

orbeckst commented on package/MDAnalysis/analysis/nucleicacids.py:
```diff
@@ -0,0 +1,187 @@
+from typing import List, Tuple, Dict, Union, NamedTuple
+
+import numpy as np
+import pandas as pd
```
avoid pandas as a dependency

ALescoulie replied:
I got rid of it in favor of a Results object

orbeckst reviewed (dismissed):
A few quick comments — I see it's in draft but this might already be helpful
- don't use pandas.DataFrame the container for the data, use our base.Results dict-like thing
- use numpydoc formatting for documentation
- eventually: update CHANGELOG and add yourself to AUTHORS

In general your idea looks good!

ojeda-e reviewed (commented):
One quick question @ALescoulie (And while this is a draft PR): is there any particular reason to use `assert_almost_equal` for your tests? I think MDAnalysis is using `assert_allclose` for float comparisons in tests. The [numpy docs](https://numpy.org/doc/stable/reference/generated/numpy.testing.assert_almost_equal.html) also suggests to use `assert_allclose` instead.

ojeda-e commented on testsuite/MDAnalysisTests/analysis/test_nucleicacids.py:
```diff
@@ -0,0 +1,66 @@
+from MDAnalysisTests.datafiles import RNA_PSF, RNA_PDB
+from numpy.testing import (
+    assert_almost_equal,
+    assert_allclose,
```
and you are importing it here already

ALescoulie replied:
I switched to `assert_all_close`, I was imitating an older set of tests in `test_nucinfo.py`

ALescoulie committed:
convert results to Results object

ALescoulie committed:
commit tests

ALescoulie commented:
I updated my code based of the feedback, I apologize for the delay, my systems programing class is keeping me really busy. I still need to write documentation for new module, but it retains the core functionality of `nucinfo.py` making measuring multiple groups and distance over time more efficient. Now I need to finish the docs.

ALescoulie commented:
@orbeckst do you want me to also update the phase and torsion functions from `nucinfo.py`

ALescoulie committed:
add nucleicacids.py to analysis imports

ALescoulie committed:
remove relative imports

ALescoulie commented:
Looking at coverage, I need to write some tests for the selection checking, It should be simple to just try passing in a protein structure and make sure it rejects the residues for not being amino acids.

orbeckst commented:
> do you want me to also update the phase and torsion functions from nucinfo.py

It's up to you. I am ok with a partial transition of nuclinfo as a start. If you don't transition all functionality, please raise individual issues for missing parts. Having small issues is good because they make for good starter issues.

orbeckst commented:
Also have a look at PEP8 — as you're starting a new module, let it start out in a PEP8 compliant state.

orbeckst commented on package/MDAnalysis/analysis/nucleicacids.py:
```diff
@@ -0,0 +1,185 @@
+
+import MDAnalysis as mda
+from ..lib import mdamath
+from ..analysis.base import AnalysisBase, Results
```
Should work as 
```suggestion
from .base import AnalysisBase, Results
```

orbeckst commented on package/MDAnalysis/analysis/nucleicacids.py:
```diff
@@ -0,0 +1,185 @@
+            # Checking number of selections and building mda.AtomGroups
+            if len(s) != 2:
+                raise ValueError("Each base pair selection given as a tuple of length 2")
+            try:
+                ag1.append(self._unv.select_atoms(s[0]))
+
+            except SelectionError:
+                raise SelectionError(f"Invalid selection in {s[0]}")
```
You might not need to catch these SelectionErrors (and thus not need to test them!) — isn't it sufficiently clear to let them happen?

ALescoulie replied:
Ok, I'll just remove that.

orbeckst commented on package/MDAnalysis/analysis/nucleicacids.py:
```diff
@@ -0,0 +1,185 @@
+            self.results[i] = np.array(self._res_dict[n])
+
+
+class WCDist(NucPairDist):
```
I'd go for more expressive names such as *WatsonCrickDistance*.

ALescoulie replied:
Ok updating that. That makes sense to update.

orbeckst commented on testsuite/MDAnalysisTests/analysis/test_nucleicacids.py:
```diff
@@ -0,0 +1,63 @@
+
+
+def test_wc_dist(u):
+    sel = [(BaseSelect('RNAA', 1), BaseSelect('RNAA', 2)),
+           (BaseSelect('RNAA', 22), BaseSelect('RNAA', 23))]
+    WC = WCDist(u, sel)
```
Is the user supposed to build selections in this way? That looks quite cumbersome. 

I think I'd prefer something where the user has to provide AtomGroups, e.g. for strand_1 and and strand_2 and the residues are matched up with zip(strand_1.residues, strand_2.residues). We can then provide a helper function to build these groups but the classes themselves don't have to much guessing baggage, except identifying the correct atoms for the distance.

ALescoulie replied:
I'm trying to get an idea of how you want this to work, so the user provides the strands, then the residue pairs are formed in a list of tuples. The classes would then take the residue selections, use the zipped list to find the specific residues than grab the specific atoms. The super class would just accept a list of AtomGroups then merge them into a single AtomGroup. I think to do this I'd write a helper function to select the strands of DNA from a Universe. Is this what you are envisioning?

The question I have it are how do you select a specific strand of strand of DNA?

ALescoulie replied:
I mostly used the named tuple for the sake of organizing inputs.

orbeckst replied:
>  so the user provides the strands, 

Yes, right now make it the user's responsibility to identify the WC pairs. `strand_1.residues[i]` will be considered base pairing with `strand_2.residues[i]`.

This is similar to the existing function, which asks the user to identify a single base pair via resid and segid. However, for your module we should move to an object-oriented approach and just work with AtomGroups instead of resid/segid selections.

> then the residue pairs are formed in a list of tuples. 

Yes.

> The classes would then take the residue selections, 

They take `strand_1.residues` and `strand_2.residues`

> use the zipped list to find the specific residues 

Yes.

> than grab the specific atoms. 

This is where the class does some actual work in order to identify the specific atoms that are required for the W-C base pair. Depending on the residue type, a different atom is chosen.

> The super class would just accept a list of AtomGroups then merge them into a single AtomGroup. 

I don't think that it would merge them into a single group. Ultimately you want to be able to use `calc_bonds()` on two arrays of positions. Each strand should have N residues and in the end, for each strand you should end up with an AtomGroup with N atoms, where the atom is either an N3 or and N1, depending on the nucleobase.

ALescoulie replied:
I was thinking it would merge them into two atom groups which then are put into calc_bonds

hmacdope commented:
@ALescoulie I love the typed code! It looks amazing. 👍 I was going to add that you need a bunch of tests for the errors you are raising. Have a look at the lines codecov is complaining about. :)

orbeckst commented on package/MDAnalysis/analysis/nucleicacids.py:
```diff
@@ -0,0 +1,185 @@
+        self._res_dict['time'] = []
+
+    def _single_frame(self) -> None:
+        dist = self._s1.positions - self._s2.positions
+        wc: np.ndarray = mdamath.pnorm(dist)
```
use [calc_bonds()](https://docs.mdanalysis.org/stable/documentation_pages/lib/distances.html?highlight=calc_bonds#MDAnalysis.lib.distances.calc_bonds) because it will properly take PBC into account

orbeckst commented:
You can reduce the scope of this PR and focus on the base class and just the WatsonCrick pair distance — the other functionality can then be added separately later. Just raise issues for the missing parts once the initial PR has been merged.

richardjgowers commented on package/MDAnalysis/analysis/nucleicacids.py:
```diff
@@ -0,0 +1,185 @@
+        sel_str: List[Tuple[str, str]] = []
+
+        for s in selections:
+            if universe.select_atoms(f" resid {s[0].resid} ").resnames[0] in ["DC", "DT", "U", "C", "T", "CYT", "THY", "URA"]:
+                a1, a2 = n3_name, n1_name
+            elif universe.select_atoms(f" resid {s[0].resid} ").resnames[0] in ["DG", "DA", "A", "G", "ADE", "GUA"]:
+                a1, a2 = n1_name, n3_name
```
Worth bringing these hardcoded names out into a kwarg so someone with different names can still use the class?

ALescoulie replied:
Ok I think a good way to do this would be a kwarg for each base. I can pretty easily set that up.

richardjgowers reviewed (changes requested):
I think @orbeckst has already said this, but I'm much more in favour of writing analysis where the atom selecting happens outside of the analysis class/function... so


```python
ag1 = u.select_atoms('foo')
ag2 = u.select_atoms('bar')

analysis(ag1, ag2)
```
rather than

```python
analysis(u, 'foo', 'bar')
```

The former is much easier to debug if you're suspicious about what's happening inside the analysis class, and also caters for people who might have differently named systems.

ALescoulie commented:
> I think @orbeckst has already said this, but I'm much more in favour of writing analysis where the atom selecting happens outside of the analysis class/function... so

I agree, the one thing is needing to select different things depending on the base pair, for example selecting N1 or N3 depending on the base pair, so I think I'll do something where they select the base pairs and my just selects the specific atoms.

richardjgowers reviewed (approved):
This is good to go as it stands.  I think as a modernisation I'd like to convert this from passing in selection strings to passing in AtomGroups, but this can be a separate step so we can enjoy the benefits of this code as-is.

ALescoulie commented:
@richardjgowers I'd like to make a few quick changes as well as add some docs then we can merge.

ALescoulie committed:
remove internal selection and simplify pr

ALescoulie committed:
add nuc acid name kwargs

ALescoulie commented:
@richardjgowers I implemented the suggestions from @orbeckst, removed selections from inside the Classes aside from selecting atoms and added keyword arguments for each base pair. I still need to add documentation but am working on that now.

ALescoulie committed:
update docs

ALescoulie commented:
ok I wrote some docs. I made sure to keep the original citation from nuclinfo in since I essentially copied her system for solving the problem, just restructured it to a faster OOP approach.

ALescoulie committed:
add analysis to tree

ALescoulie committed:
fix docs

ALescoulie committed:
update AUTHORS and CHANGELOG

ALescoulie commented:
I got the docs working and moved the selection outside the classes

ALescoulie committed:
fix username in CHANGELOG

richardjgowers committed:
Merge branch 'develop' into wc_pair_perf
